src/credentials/FabricTable.cpp

```diff
@@ -787,8 +816,230 @@ void FabricTable::DeleteAllFabrics()
+    const auto length = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(length), CHIP_ERROR_BUFFER_TOO_SMALL);
+    DefaultStorageKeyAllocator keyAlloc;
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(keyAlloc.LastKnownGoodTimeKey(), buf, static_cast<uint16_t>(length)));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FabricTable::StoreLastKnownGoodChipEpochTime(System::Clock::Seconds32 lastKnownGoodChipEpochTime) const
+{
+    return StoreLastKnownGoodChipEpochTime(lastKnownGoodChipEpochTime, NullOptional);
+}
+
+CHIP_ERROR FabricTable::InitLastKnownGoodChipEpochTime()
+{
+    // 3.5.6.1 Last Known Good UTC Time:
+    //
+    // "A Node’s initial out-of-box Last Known Good UTC time SHALL be the
+    // compile-time of the firmware."
+    //
+    // Read build time from the configuration manager to get our initial Last
+    // Known Good Time.  Note 'SHALL' above: it is impermissible for a platform
+    // not to know this.  Hence return an error to the caller if build time is
+    // not available.  Callers should consider this fatal.
+    System::Clock::Seconds32 buildTime;
+    ReturnErrorOnFailure(DeviceLayer::ConfigurationMgr().GetFirmwareBuildChipEpochTime(buildTime));
+    System::Clock::Seconds32 kvstoreLastKnownGoodChipEpochTime;
+    CHIP_ERROR err = LoadLastKnownGoodChipEpochTime(kvstoreLastKnownGoodChipEpochTime);
+    VerifyOrReturnError(err == CHIP_NO_ERROR || err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND, err);
+    ChipLogProgress(Discovery, "Last Known Good Time: %s",
+                    err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND ? "[unknown]"
+                                                                        : FormatChipEpochTime(kvstoreLastKnownGoodChipEpochTime));
+    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND || buildTime > kvstoreLastKnownGoodChipEpochTime)
+    {
+        // If we have no value in persistence, or the firmware build time is
+        // later than the value in persistence, set last known good time to the
+        // firmware build time and write back.
+        ChipLogProgress(Discovery, "Setting Last Known Good Time to firmware build time %s", FormatChipEpochTime(buildTime));
+        mLastKnownGoodChipEpochTime.SetValue(buildTime);
+        return StoreLastKnownGoodChipEpochTime(buildTime);
+    }
+    else
+    {
+        mLastKnownGoodChipEpochTime.SetValue(kvstoreLastKnownGoodChipEpochTime);
+        return CHIP_NO_ERROR;
+    }
+}
+
+CHIP_ERROR FabricTable::SetLastKnownGoodChipEpochTime(System::Clock::Seconds32 lastKnownGoodChipEpochTime)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mLastKnownGoodChipEpochTime.HasValue(), err = CHIP_ERROR_INCORRECT_STATE);
+    ChipLogProgress(Discovery, "Last Known Good Time: %s", FormatChipEpochTime(mLastKnownGoodChipEpochTime.Value()));
+    ChipLogProgress(Discovery, "new proposed Last Known Good Time: %s", FormatChipEpochTime(lastKnownGoodChipEpochTime));
```

Comment:
All your logging tags are `Discovery` but this code does not relate to discovery

Reply:
I was copying the existing convention in FabricTable.  Is there a better log module to use?